src/Features/Core/Portable/PullMemberUp/AbstractPullMemberUpRefactoringProvider.cs

```diff
@@ -0,0 +1,99 @@
+            _cancellationToken = context.CancellationToken;
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var userSelectedNode = root.FindNode(context.Span);
+
+
+            if (!IsSelectionValid(context.Span, userSelectedNode))
+            {
+                return;
+            }
+
+            var userSelectNodeSymbol = semanticModel.GetDeclaredSymbol(userSelectedNode);
+            if (userSelectNodeSymbol == null || userSelectNodeSymbol.ContainingType == null)
+            {
+                return;
+            }
+
+            var allTargets = FindAllValidTargets(userSelectNodeSymbol.ContainingType, context.Document.Project.Solution);
+
+            if (allTargets.Length == 0)
+            {
+                return;
+            }
+
+            if ((userSelectNodeSymbol is IMethodSymbol methodSymbol && methodSymbol.IsOrdinaryMethod()) ||
+                userSelectNodeSymbol.Kind == SymbolKind.Property ||
+                userSelectNodeSymbol.Kind == SymbolKind.Event ||
+                userSelectNodeSymbol.Kind == SymbolKind.Field)
+            {
+                await PullMemberUpViaQuickAction(context, userSelectNodeSymbol, allTargets);
+            }
+        }
+
+        private ImmutableArray<INamedTypeSymbol> FindAllValidTargets(
+            INamedTypeSymbol selectedNodeOwnerSymbol,
+            Solution solution)
+        {
+            return selectedNodeOwnerSymbol.AllInterfaces.Concat(selectedNodeOwnerSymbol.GetBaseTypes()).
+                Where(baseType => baseType != null &&
+                    baseType.DeclaringSyntaxReferences.Length > 0 &&
+                    IsLocationValid(baseType, solution)).ToImmutableArray();
+        }
+
+        private bool IsLocationValid(INamedTypeSymbol symbol, Solution solution)
+        {
+            return symbol.Locations.Any(location => location.IsInSource &&
+                !solution.GetDocument(location.SourceTree).IsGeneratedCode(_cancellationToken));
+        }
+
+        private async Task PullMemberUpViaQuickAction(
+            CodeRefactoringContext context,
+            ISymbol userSelectNodeSymbol,
+            ImmutableArray<INamedTypeSymbol> targets)
+        {
+            foreach (var target in targets)
+            {
+                AbstractMemberPullerWithQuickAction puller = default;
+                if (target.TypeKind == TypeKind.Interface &&
+                    userSelectNodeSymbol.Kind != SymbolKind.Field)
+                {
+                    puller = new InterfacePullerWithQuickAction();
+                }
+                else
+                {
+                    puller = new ClassPullerWithQuickAction();
+                }
+
+                var action = await puller.TryComputeRefactoring(context.Document, userSelectNodeSymbol, target);
```

Comment:
is this action lazy? or are you computing it up front?

Reply:
This action is lazy, But before create it there is still some checks need to be done. Some front part of this PR could be put into the lazy part. I will update them in next commit.